json_patch.c

```diff
@@ -0,0 +1,249 @@
+		json_object_put(*res);
+		*res = NULL;
+		errno = ENOENT;
+		return -1;
+	}
+
+	return 0;
+}
+
+static int __json_patch_apply_remove(struct json_pointer_get_result *jpres)
+{
+	if (json_object_is_type(jpres->parent, json_type_array)) {
+		return json_object_array_del_idx(jpres->parent, jpres->id.index, 1);
+	} else if (jpres->parent && jpres->id.key) {
+		json_object_object_del(jpres->parent, jpres->id.key);
+		return 0;
+	} else {
+		return json_object_put(jpres->obj);
+	}
+}
+
+static int json_patch_apply_remove(struct json_object **res, const char *path)
+{
+	struct json_pointer_get_result jpres;
+
+	if (json_pointer_get_internal(*res, path, &jpres))
+		return -1;
+
+	return __json_patch_apply_remove(&jpres);
+}
+
+static int json_object_array_insert_idx_cb(struct json_object *parent, size_t idx,
+                                           struct json_object *value, void *priv)
+{
+	int *add = priv;
+
+	if (idx > json_object_array_length(parent))
+	{
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (*add)
+		return json_object_array_insert_idx(parent, idx, value);
+	else
+		return json_object_array_put_idx(parent, idx, value);
+}
+
+static int json_patch_apply_add_replace(struct json_object **res,
+                                        struct json_object *patch_elem,
+                                        const char *path, int add)
+{
+	struct json_object *value;
+	int rc;
+
+	if (!json_object_object_get_ex(patch_elem, "value", &value)) {
+		errno = EINVAL;
+		return -1;
+	}
+	/* if this is a replace op, then we need to make sure it exists before replacing */
+	if (!add && json_pointer_get(*res, path, NULL)) {
+		errno = ENOENT;
+		return -1;
+	}
+
+	rc = json_pointer_set_with_array_cb(res, path, json_object_get(value),
+					    json_object_array_insert_idx_cb, &add);
+	if (rc)
+		json_object_put(value);
+
+	return rc;
+}
+
+static int json_object_array_move_cb(struct json_object *parent, size_t idx,
+                                     struct json_object *value, void *priv)
+{
+	struct json_pointer_get_result *from = priv;
+	size_t len = json_object_array_length(parent);
+
+	/**
+	 * If it's the same array parent, it means that we removed
+	 * and element from it, so the length is temporarily reduced
+	 * by 1, which means that if we try to move an element to
+	 * the last position, we need to check the current length + 1
+	 */
+	if (parent == from->parent)
+		len++;
+
+	if (idx > len)
+	{
+		errno = EINVAL;
+		return -1;
+	}
+
+	return json_object_array_insert_idx(parent, idx, value);
+}
+
+static int json_patch_apply_move_copy(struct json_object **res,
+                                      struct json_object *patch_elem,
+                                      const char *path, int move)
+{
+	json_pointer_array_set_cb array_set_cb;
+	struct json_pointer_get_result from;
+	struct json_object *jfrom;
+	const char *from_s;
+	size_t from_s_len;
+	int rc;
+
+	if (!json_object_object_get_ex(patch_elem, "from", &jfrom)) {
+		errno = EINVAL;
```

Comment:
Let's avoid (ab)using errno to return additional json-c-specific error information.  Instead, consider calling _json_c_set_last_err() with a more detailed message.

Reply:
Also, it would be worthwhile adding a json patch specific error code param to the json_patch_apply() signature, to ensure it's possible for callers to implement the recommended [error handling](https://tools.ietf.org/html/rfc5789#section-2.2) based on how things actually failed.

Reply:
I'm working on this in my fork

Reply:
I added a `struct json_patch_error` arg to `json_patch_apply()`.